### PR TITLE
pass the basic test of cmd_splitter

### PR DESCRIPTION
**Pull request by hetong07**
very init version

### CODE REVIEW

File: src/drive/frontend/cmd_splitter.cpp [hunk 1]
```diff
@@ -1,15 +1,34 @@
 #include "cmd.h"
 #include "staccel_kernel.h"
 
-void cmd_splitter(ST_Queue<SuperCommand> super_command_queue,
-		  ST_Queue<Command> command_queue) {
+
```
Comment (zainryan) [inline src/drive/frontend/cmd_splitter.cpp]: Remove these unnecessary empty lines.

File: src/drive/frontend/cmd_splitter.cpp [hunk 2]
```diff
@@ -1,15 +1,34 @@
...
+  if (super_command_queue.read_nb(super_command)) {
+    Command command;
+    for (int i = 0; i < SUPERCOMMAND_SIZE_IN_BIT; i += COMMAND_SIZE_IN_BIT) {
+      command.index = 
+        super_command.data(i + COMMAND_SIZE_IN_BIT - 1, 
+            i + COMMAND_NUM_SIZE_IN_BIT + COMMAND_ISREAD_SIZE_IN_BIT);
+      command.num = super_command.data(i + COMMAND_NUM_SIZE_IN_BIT + COMMAND_ISREAD_SIZE_IN_BIT - 1,
+          i + COMMAND_ISREAD_SIZE_IN_BIT);
+      command.is_read = super_command.data(i + COMMAND_ISREAD_SIZE_IN_BIT - 1, i);
+      command_queue.write_nb(command);
```
Comment (zainryan) [inline src/drive/frontend/cmd_splitter.cpp]: Use blocking write.

File: src/drive/frontend/cmd_splitter.cpp [hunk 3]
```diff
@@ -1,15 +1,34 @@
...
+#define COMMAND_NUM_SIZE_IN_BIT 13
+#define COMMAND_ISREAD_SIZE_IN_BIT 1
+
+inline void cmd_splitter_impl(ST_Queue<SuperCommand> &super_command_queue,
+    ST_Queue<Command> &command_queue) {
+  SuperCommand super_command;
+  if (super_command_queue.read_nb(super_command)) {
+    Command command;
+    for (int i = 0; i < SUPERCOMMAND_SIZE_IN_BIT; i += COMMAND_SIZE_IN_BIT) {
+      command.index =
```
Comment (zainryan) [inline src/drive/frontend/cmd_splitter.cpp]: Add #pragma HLS unroll here.

File: src/drive/frontend/cmd_splitter.hpp [hunk 4]
```diff
@@ -0,0 +1,4 @@
+#pragma once
+void cmd_splitter(ST_Queue<SuperCommand> &super_command_queue, ST_Queue<Command> &command_queue);
+
```
Comment (zainryan) [inline src/drive/frontend/cmd_splitter.hpp]: No empty line.

File: test/drive/frontend/CMakeLists.txt [hunk 5]
```diff
@@ -14,4 +14,13 @@ add_test(
...
\ No newline at end of file
+  )
+
+target_include_directories(
+    unit_tests
+    PUBLIC
+    ${CMAKE_SOURCE_DIR}/src/drive/frontend/
+    ${CMAKE_SOURCE_DIR}/third_party
+)
+
```
Comment (zainryan) [inline test/drive/frontend/CMakeLists.txt]: empty lines.

File: test/drive/frontend/cmd_splitter.cpp [hunk 6]
```diff
@@ -1,5 +1,74 @@
 #include "gtest/gtest.h"
+#include "cmd.h"
+#include "stdlib.h"
+#include "staccel_type.h"
```
Comment (zainryan) [inline test/drive/frontend/cmd_splitter.cpp]: sort headers alphabetically.

File: test/drive/frontend/cmd_splitter.cpp [hunk 7]
```diff
@@ -1,5 +1,74 @@
...
+#include <thread>
+#include <vector>
+#include "cmd_splitter.hpp"
+
+#define COMMAND_SIZE_IN_BIT 64
+#define MAX_READ_TIMES  100000000L
```
Comment (zainryan) [inline test/drive/frontend/cmd_splitter.cpp]: two spaces -> one space

File: test/drive/frontend/cmd_splitter.cpp [hunk 8]
```diff
@@ -1,5 +1,74 @@
...
+#define NUM_OF_COMMAND 8
+
+inline void command_eq(Command &expected, Command &real) {
+  ASSERT_EQ(expected.index, real.index);
+  ASSERT_EQ(expected.num, real.num);
+  ASSERT_EQ(expected.is_read, real.is_read);
+}
+
```
Comment (zainryan) [inline test/drive/frontend/cmd_splitter.cpp]: empty line

File: test/drive/frontend/cmd_splitter.cpp [hunk 9]
```diff
@@ -1,5 +1,74 @@
...
+#include "cmd.h"
+#include "stdlib.h"
+#include "staccel_type.h"
+#include <thread>
+#include <vector>
+#include "cmd_splitter.hpp"
+
+#define COMMAND_SIZE_IN_BIT 64
+#define MAX_READ_TIMES  100000000L
+#define NUM_OF_COMMAND 8
```
Comment (zainryan) [inline test/drive/frontend/cmd_splitter.cpp]: Should be calculated using the defined macros in cmd_splitter.cpp

File: test/drive/frontend/cmd_splitter.cpp [hunk 10]
```diff
@@ -1,5 +1,74 @@
...
+
+TEST(test_cmd_splitter, correct_split) {
+
+  std::vector<Command> expected;
+  std::vector<Command> real;
+
+  SuperCommand super_command;
+  Command command;
+
+  for (int i = 0; i < NUM_OF_COMMAND; i ++) {
```
Comment (zainryan) [inline test/drive/frontend/cmd_splitter.cpp]: i ++ -> i++

File: test/drive/frontend/cmd_splitter.cpp [hunk 11]
```diff
@@ -1,5 +1,74 @@
...
+    int upper_half_command_index = rand();
+    int lower_half_command_index = rand();
+    int command_num = rand();
+    bool command_is_read = (bool) (rand() % 2);
+    command.index = 
+      ap_uint<50> (((long long) upper_half_command_index) << 32 | lower_half_command_index);
+    command.num = (ap_uint<13>) command_num;
```
Comment (zainryan) [inline test/drive/frontend/cmd_splitter.cpp]: Use defined macros in  cmd_splitter.cpp

File: test/drive/frontend/cmd_splitter.cpp [hunk 12]
```diff
@@ -1,5 +1,74 @@
...
+    bool command_is_read = (bool) (rand() % 2);
+    command.index = 
+      ap_uint<50> (((long long) upper_half_command_index) << 32 | lower_half_command_index);
+    command.num = (ap_uint<13>) command_num; 
+    command.is_read = command_is_read; 
+    expected.push_back(command);
+
+    long long command_content = 0;
+    command_content |= command.index;
+    command_content <<= 13;
```
Comment (zainryan) [inline test/drive/frontend/cmd_splitter.cpp]: Use defined macros in cmd_splitter.cpp

File: test/drive/frontend/cmd_splitter.cpp [hunk 13]
```diff
@@ -1,5 +1,74 @@
...
+    command_content |= command.num;
+    command_content <<= 1;
```
Comment (zainryan) [inline test/drive/frontend/cmd_splitter.cpp]: ditto

File: test/drive/frontend/cmd_splitter.cpp [hunk 14]
```diff
@@ -1,5 +1,74 @@
...
+    command_content |= command.is_read;
+
+    super_command.data <<= 64;
```
Comment (zainryan) [inline test/drive/frontend/cmd_splitter.cpp]: ditto

File: test/drive/frontend/cmd_splitter.cpp [hunk 15]
```diff
@@ -1,5 +1,74 @@
...
+  cmd_splitter_impl(super_command_queue, command_queue);
+
+  // read command from the command queue
+  int read_times = 0;
+  while (real.size() != expected.size() && read_times < MAX_READ_TIMES) {
+    Command command;
+    if (command_queue.read_nb(command)) {
+      real.push_back(command); 
+    }
+    read_times ++;
```
Comment (zainryan) [inline test/drive/frontend/cmd_splitter.cpp]:
```suggestion
    read_times++;
```

File: test/drive/frontend/cmd_splitter.cpp (hunk 7 above)
Comment (zainryan) [reply test/drive/frontend/cmd_splitter.cpp]: This should not be defined here. I suggest that we define it in some global header for the test folder.

File: test/drive/frontend/cmd_splitter.cpp [hunk 16]
```diff
@@ -1,5 +1,74 @@
...
+    super_command.data |= ((ap_uint<64>) command_content);
+  }
+
+  ST_Queue<SuperCommand> super_command_queue(8);
+  ST_Queue<Command> command_queue(32); 
+  
+  // put super-command to super command queue
+  ASSERT_EQ(super_command_queue.write_nb(super_command), true); 
+
+  cmd_splitter_impl(super_command_queue, command_queue);
```
Comment (zainryan) [inline test/drive/frontend/cmd_splitter.cpp]: Let's define a macro in some global header for the test folder.

#define RUN_METHOD(times, method) for (int i = 0; i < times; i++) method

And replace this line with RUN_METHOD(1, cmd_splitter_impl(super_command_queue, command_queue))

File: test/drive/frontend/cmd_splitter.cpp [hunk 17]
```diff
@@ -1,5 +1,80 @@
...
+  ST_Queue<SuperCommand> super_command_queue(8);
+  ST_Queue<Command> command_queue(32); 
+
+  // put super-command to super command queue
+  ASSERT_EQ(super_command_queue.write_nb(super_command), true); 
+
+  RUN_METHOD(1, cmd_splitter_impl(super_command_queue, command_queue));
+
+  // read command from the command queue
+  int read_times = 0;
```
Comment (zainryan) [inline test/drive/frontend/cmd_splitter.cpp]: Define a template method.

template <typename T> void drain_queue(ST_Queue<T> &queue, std::vector<T> &vec);
It drains the queue (for DRAIN_QUEUE_READ_TIMES times) and puts the results into the vector.

File: test/drive/frontend/cmd_splitter.cpp [hunk 18]
```diff
@@ -1,5 +1,80 @@
...
+  while (real.size() != expected.size() && read_times < MAX_READ_TIMES) {
```
Comment (zainryan) [inline test/drive/frontend/cmd_splitter.cpp]: Change MAX_READ_TIMES into DRAIN_QUEUE_READ_TIMES

File: test/drive/frontend/cmd_splitter.cpp (hunk 17 above)
Comment (zainryan) [reply test/drive/frontend/cmd_splitter.cpp]: Sorry, it should be template void drain_queue(ST_Queue *queue, std::vector *vec);

File: src/drive/frontend/cmd_splitter.hpp [hunk 19]
```diff
@@ -0,0 +1,4 @@
+#pragma once
+#include "staccel_type.h"
+void cmd_splitter(ST_Queue<SuperCommand> &super_command_queue, ST_Queue<Command> &command_queue);
```
[the remark this answers is not in the document]
Comment (zainryan) [reply src/drive/frontend/cmd_splitter.hpp]: Should be void cmd_splitter(ST_Queue<SuperCommand> *super_command_queue, ST_Queue<Command> *command_queue);

File: src/drive/frontend/cmd_splitter.cpp [hunk 20]
```diff
@@ -1,15 +1,28 @@
 #include "cmd.h"
+#include "cmd_splitter.hpp"
 #include "staccel_kernel.h"
 
-void cmd_splitter(ST_Queue<SuperCommand> super_command_queue,
-		  ST_Queue<Command> command_queue) {
+inline void cmd_splitter_impl(ST_Queue<SuperCommand> *super_command_queue,
+    ST_Queue<Command> *command_queue) {
+  SuperCommand super_command;
+  if (super_command_queue -> read_nb(super_command)) {
```
Comment (zainryan) [inline src/drive/frontend/cmd_splitter.cpp]: no space around ->

File: test/drive/frontend/cmd_splitter.cpp [hunk 21]
```diff
@@ -1,5 +1,73 @@
...
+  ST_Queue<Command> command_queue(32); 
+
+  // put super-command to super command queue
+  ASSERT_EQ(super_command_queue.write_nb(super_command), true); 
+
+  RUN_METHOD(1, cmd_splitter_impl(&super_command_queue, &command_queue));
+
+  // read command from the command queue
+  drain_queue(&command_queue, &real);
+  //int read_times = 0;
```
Comment (zainryan) [inline test/drive/frontend/cmd_splitter.cpp]: delete comments.

File: test/inc/test_utils.hpp [hunk 22]
```diff
@@ -0,0 +1,18 @@
...
+#include "test_params.hpp"
+
+#define RUN_METHOD(times, method) for(int i = 0; i < times; i++) method
+
+template <typename T> 
+void drain_queue(ST_Queue<T> *queue, std::vector<T> *vec) {
+  unsigned long long read_times = 0;
+  while (read_times < DRAIN_QUEUE_READ_TIMES) {
+    T data;
+    if (queue -> read_nb(data)) {
```
Comment (zainryan) [inline test/inc/test_utils.hpp]: ->

File: test/inc/test_utils.hpp [hunk 23]
```diff
@@ -0,0 +1,18 @@
...
+      (*vec).push_back(data);
```
Comment (zainryan) [inline test/inc/test_utils.hpp]: vec->

File: src/drive/frontend/cmd_splitter.cpp [hunk 24]
```diff
@@ -1,15 +1,28 @@
...
+    Command command;
+    for (int i = 0; i < SUPERCOMMAND_SIZE_IN_BIT; i += COMMAND_SIZE_IN_BIT) {
+#pragma HLS unroll
+      command.index = 
+        super_command.data(i + COMMAND_SIZE_IN_BIT - 1, 
+            i + COMMAND_NUM_SIZE_IN_BIT + COMMAND_ISREAD_SIZE_IN_BIT);
+      command.num = super_command.data(i + COMMAND_NUM_SIZE_IN_BIT + COMMAND_ISREAD_SIZE_IN_BIT - 1,
+          i + COMMAND_ISREAD_SIZE_IN_BIT);
+      command.is_read = super_command.data(i + COMMAND_ISREAD_SIZE_IN_BIT - 1, i);
+      command_queue -> write(command);
```
Comment (zainryan) [inline src/drive/frontend/cmd_splitter.cpp]: ->

**Comment by hetong07** [comment]
Resolve #15

**Pull request merged**